### PR TITLE
OD-14169 Documents API: Generating file's preview

### PR DESCRIPTION
**Pull request by KeyrisXdSnow**
… .XML and other text formats

### CODE REVIEW

**Review by dlitvinko — DISMISSED** [review]

File: server/src/main/groovy/ish/oncourse/server/api/v1/function/DocumentFunctions.groovy [hunk 1]
```diff
@@ -180,13 +180,34 @@ class DocumentFunctions {
...
+                        version.thumbnail = ThumbnailGenerator.generateForExcel(content)
+                    } else {
+                        if (isCsv(version.mimeType)) {
+                            version.thumbnail = ThumbnailGenerator.generateForCsv(content)
+                        } else {
+                            if (isText(version.mimeType)) {
+                                version.thumbnail = ThumbnailGenerator.generateForText(content)
+                            } else {
+                                version.thumbnail = generatePdfPreview(content)
+                            }
```
Comment (dmitars) [inline server/src/main/groovy/ish/oncourse/server/api/v1/function/DocumentFunctions.groovy:203]: What about factory / factory method / enums? number of spaces looks ambiguous. Also combination of else{..} blocks, which contain if{...}else{...} operator is archaic. Why don't use else if{...} / switch ?

Comment (dmitars) [inline server/src/main/groovy/ish/oncourse/server/api/v1/function/DocumentFunctions.groovy:203]: What about factory / factory method / enums? number of spaces looks ambiguous. Also combination of else{..} blocks, which contain if{...}else{...} operator is archaic. Why don't use else if{...} / switch ?

File: server/src/main/groovy/ish/oncourse/server/api/v1/service/impl/ExportTemplateApiImpl.groovy [hunk 2]
```diff
@@ -31,6 +31,12 @@ class ExportTemplateApiImpl implements ExportTemplateApi {
         service.get(id)
     }
 
+    @Override
+    byte[] getHighQualityPreview(Long id) {
+        def content = service.getPreview(id)
+        return content
```
Comment (dmitars) [inline server/src/main/groovy/ish/oncourse/server/api/v1/service/impl/ExportTemplateApiImpl.groovy:37]: Why do we need content variable?

File: server/src/main/groovy/ish/oncourse/server/export/ExportService.groovy [hunk 3]
```diff
@@ -159,4 +165,16 @@ class ExportService {
...
 	}
+
+	static void createExportPreview(ExportTemplate template, byte[] content) {
+		if (FilenameUtils.getExtension(template.keyCode).equalsIgnoreCase("csv")) {
+			template.preview = ThumbnailGenerator.generateForCsv(content)
+			template.objectContext.commitChanges()
+		} else {
+			if (FilenameUtils.getExtension(template.keyCode).equalsIgnoreCase("xml")) {
+				template.preview = ThumbnailGenerator.generateForText(content)
+				template.objectContext.commitChanges()
```
Comment (dmitars) [inline server/src/main/groovy/ish/oncourse/server/export/ExportService.groovy:176]: FilenameUtils.getExtentions(...) and template.objectContext.commitChanges() are duplicates

File: server/src/main/java/ish/oncourse/server/api/service/ExportTemplateApiService.java [hunk 4]
```diff
@@ -63,6 +66,10 @@ public byte[] exportOnDisk(Long id) {
         return getEntityAndValidateExistence(cayenneService.getNewContext(), id).getScript().getBytes();
     }
 
+    public byte[] getPreview(Long id){
+        return ObjectSelect.columnQuery(ExportTemplate.class, Report.PREVIEW).where(Report.ID.eq(id)).selectOne(cayenneService.getNewContext());
+    }
```
Comment (dmitars) [inline server/src/main/java/ish/oncourse/server/api/service/ExportTemplateApiService.java:71]: It works, but it is better to use ExportTemplate properties for ExportTemplate queries

File: server/src/main/java/ish/util/ImageHelper.java [hunk 5]
```diff
@@ -188,34 +187,57 @@ public static boolean isImage(byte[] content, String mimeType) {
...
-	 * @return - binary content of generated thumbnail, null - if transformation can't be performed
+	 * Determines if document version is a Microsoft Office document by its MIME type
+	 * according <a href="https://www.iana.org/assignments/media-types/media-types.xhtml"> IANA </a>}
 	 */
-	public static byte[] generateThumbnail(byte[] content, String type) throws IOException {
-		String format = type;
+	public static boolean isDoc(String mimeType) {
+		return mimeType.equals("application/msword") ||
+				mimeType.equals("application/vnd.openxmlformats-officedocument.wordprocessingml.document");
+	}
```
Comment (dmitars) [inline server/src/main/java/ish/util/ImageHelper.java:196]: Different code styles for different mime types in methods

Also it is better don't use set of constants like graph of management flags, even hardcoded lists are more safe

File: server/src/main/java/ish/util/ImageHelper.java [hunk 6]
```diff
@@ -188,34 +187,57 @@ public static boolean isImage(byte[] content, String mimeType) {
...
+			case "text/calendar":
+			case "text/javascript":
+			case "application/vnd.oasis.opendocument.text":
+			case "application/x-httpd-php":
+			case "text/rtf":
+			case "text/plain":
+			case "text/xml":
+				return true;
+			default:
+				return false;
```
Comment (dmitars) [inline server/src/main/java/ish/util/ImageHelper.java:237]: return docTypes.contains(mimeType)

File: server/src/main/java/ish/util/ThumbnailGenerator.groovy [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+            image = image.getSubimage(0, 0, image.width, image.height - 5)
+            def scaledSize = calculateThumbnailSize(image, true)
+
+            ByteArrayOutputStream scaledImageOutput = new ByteArrayOutputStream()
+            Thumbnails.of(new ByteArrayInputStream(getAsByteArray(image, PDF_PREVIEW_FORMAT)))
+                    .size(scaledSize.weight, scaledSize.height)
+                    .outputFormat(PDF_PREVIEW_FORMAT)
+                    .toOutputStream(scaledImageOutput)
+
+            return scaledImageOutput.toByteArray()
```
Comment (dmitars) [inline server/src/main/java/ish/util/ThumbnailGenerator.groovy:130]: seems like partly duplicated with csv

**Review by dmitars — APPROVED** [review]

**Review by yuryyas — DISMISSED** [review]

File: server/src/main/java/ish/oncourse/server/api/service/ExportTemplateApiService.java (hunk 4 above)
Comment (KeyrisXdSnow) [reply server/src/main/java/ish/oncourse/server/api/service/ExportTemplateApiService.java:71]: Fixed

File: server/src/main/java/ish/util/ImageHelper.java (hunk 5 above)
Comment (KeyrisXdSnow) [reply server/src/main/java/ish/util/ImageHelper.java:198]: Fixed

File: server/src/main/groovy/ish/oncourse/server/api/v1/function/DocumentFunctions.groovy (hunk 1 above)
Comment (KeyrisXdSnow) [reply server/src/main/groovy/ish/oncourse/server/api/v1/function/DocumentFunctions.groovy]: Constructions made as switch

**Review by GFilipovich — APPROVED** [review]

**Pull request merged**